entry.go

```diff
@@ -197,27 +210,167 @@ func (te *TunaEntry) StartReverse(stream *smux.Stream) error {
+	var npc *nknsdk.NanoPayClaimer
+	lastComputed := common.Fixed64(0)
+	lastClaimed := common.Fixed64(0)
+	lastUpdate := time.Now()
+	isClosed := false
+
+	npc, err = te.wallet.NewNanoPayClaimer(int32(claimInterval/time.Millisecond), onErr, te.config.BeneficiaryAddr)
+	if err != nil {
+		log.Fatalln(err)
+	}
 
 	go func() {
-		session, err := te.getSession(false)
-		if err != nil {
-			return
-		}
-		stream, err := session.OpenStream()
-		if err != nil {
-			return
-		}
-		stream.Close()
 		for {
-			_, err = session.AcceptStream()
+			err := <-onErr.C
 			if err != nil {
-				log.Println("Close connection:", err)
-				te.close()
-				return
+				log.Println("Couldn't claim nano pay:", err)
+				if npc.IsClosed() {
+					Close(session)
+					Close(conn)
+					isClosed = true
+					break
+				}
 			}
 		}
 	}()
 
+	go func() {
+		for {
+			time.Sleep(claimInterval)
+
+			if isClosed {
+				break
+			}
+
+			if time.Since(lastUpdate) > claimInterval {
+				log.Println("Didn't update nano pay for more than", claimInterval.String())
+				Close(session)
+				Close(conn)
+				isClosed = true
+				break
+			}
+
+			if common.Fixed64(float64(lastComputed)*0.9) > lastClaimed {
+				log.Println("Nano pay amount covers less than 90% of total cost")
+				Close(session)
+				Close(conn)
+				isClosed = true
+				break
+			}
+		}
+	}()
+
+	for {
+		stream, err := session.AcceptStream()
+		if err != nil {
+			log.Println("Couldn't accept stream:", err)
+			break
+		}
+
+		metadata := stream.Metadata()
+		if len(metadata) == 0 { // payment stream
+			if te.config.Reverse {
+				continue
+			}
+			go func(stream *smux.Stream) {
+				txData, err := ioutil.ReadAll(stream)
+				if err != nil && err.Error() != io.EOF.Error() {
+					log.Println("Couldn't read payment stream:", err)
+					return
+				}
+				if len(txData) == 0 {
+					return
+				}
+				tx := new(transaction.Transaction)
+				if err := tx.Unmarshal(txData); err != nil {
+					log.Println("Couldn't unmarshal payment stream data:", err)
+					return
+				}
+				amount, err := npc.Claim(tx)
+				if err != nil {
+					log.Println("Couldn't accept nano pay update:", err)
+					return
+				}
+
+				totalCost := common.Fixed64(0)
+				for i := range bytesIn {
+					in := atomic.LoadUint64(&bytesIn[i])
+					out := atomic.LoadUint64(&bytesOut[i])
+					if in == 0 && out == 0 {
+						continue
+					}
+					service, err := te.getService(byte(i))
+					if err != nil {
+						continue
+					}
+					serviceInfo := te.config.Services[service.Name]
+					entryToExitPrice, exitToEntryPrice, err := ParsePrice(serviceInfo.Price)
+					if err != nil {
+						continue
+					}
+					totalCost += entryToExitPrice*common.Fixed64(in)/TrafficUnit + exitToEntryPrice*common.Fixed64(out)/TrafficUnit
+					fmt.Println("totalCost:", totalCost)
+				}
+
+				lastComputed = totalCost
+				lastClaimed = amount.ToFixed64()
+				lastUpdate = time.Now()
+			}(stream)
+			continue
+		}
+
+		serviceId := metadata[0]
```

Comment:
All these code you copied are for exit, not for entry. The reverse mode is already functioning, only payment needs to be added. These part of the code are all irrelevant.